src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/AceEditorMonitor.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * AceEditorMonitor.java
+ *
+ * Copyright (C) 2009-19 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+package org.rstudio.studio.client.workbench.views.source.editors.text;
+
+import org.rstudio.core.client.BrowseCap;
+import org.rstudio.core.client.dom.WindowEx;
+
+import com.google.gwt.core.client.Scheduler;
+import com.google.gwt.event.logical.shared.AttachEvent;
+
+public class AceEditorMonitor
+{
+   public AceEditorMonitor(AceEditor editor)
+   {
+      editor_ = editor;
+      init();
+   }
+   
+   private void init()
+   {
+      editor_.addAttachHandler((AttachEvent event) -> {
+         
+         if (event.isAttached())
+         {
+            beginMonitoring();
+         }
+         else
+         {
+            endMonitoring();
+         }
+      });
+   }
+   
+   private boolean monitor()
+   {
+      if (!monitoring_)
+         return false;
+      
+      if (BrowseCap.isWindowsDesktop())
+      {
+         double ratio = WindowEx.get().getDevicePixelRatio();
+         editor_.setScrollSpeed(ACE_EDITOR_DEFAULT_SCROLL_SPEED / ratio);
```

Comment:
Is there any chance at all this could be expensive? (wondering if we should avoid calling it if the device pixel ratio hasn't changed in the last 1s)

Reply:
Good idea. Looking in a bit deeper, Ace allows for arbitrary 'onchange' handlers to be called in response to changes in options. While I don't think any are registered for `scrollSpeed`, it's better safe than sorry.